File: app_pojav_zh/src/main/java/com/movtery/customcontrols/ControlInfoData.java
package com.movtery.customcontrols;

public class ControlInfoData implements Comparable<ControlInfoData> {
    public String fileName;
    public String name = "null";
    public String version = "null";
    public String author = "null";
    public String desc = "null";

    public ControlInfoData() {
    }

    @Override
    public int compareTo(ControlInfoData o) {
        if (o == null) {
            throw new NullPointerException("Cannot compare to null.");
        }

        String thisName = (this.fileName != null) ? this.fileName : this.name;
        String otherName = (o.fileName != null) ? o.fileName : o.name;

        return compareChar(thisName, otherName);
    }

    private int compareChar(String first, String second) {
        int firstLength = first.length();
        int secondLength = second.length();

        //遍历两个字符串的字符
        for (int i = 0; i < Math.min(firstLength, secondLength); i++) {
            char firstChar = Character.toLowerCase(first.charAt(i));
            char secondChar = Character.toLowerCase(second.charAt(i));

            int compare = Character.compare(firstChar, secondChar);
            if (compare != 0) {
                return compare;
            }
        }

        return Integer.compare(firstLength, secondLength);
    }
}

File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/subassembly/customcontrols/ControlInfoData.kt
package com.movtery.pojavzh.ui.subassembly.customcontrols

import kotlin.math.min

class ControlInfoData : Comparable<ControlInfoData?> {
    @JvmField
    var fileName: String? = null
    @JvmField
    var name: String = "null"
    @JvmField
    var version: String = "null"
    @JvmField
    var author: String = "null"
    @JvmField
    var desc: String = "null"

    override fun compareTo(other: ControlInfoData?): Int {
        other ?: run { throw NullPointerException("Cannot compare to null.") }

        val thisName = this.fileName ?: this.name
        val otherName = other.fileName ?: other.name

        return compareChar(thisName, otherName)
    }

    private fun compareChar(first: String?, second: String?): Int {
        val firstLength = first!!.length
        val secondLength = second!!.length

        //遍历两个字符串的字符
        for (i in 0 until min(firstLength.toDouble(), secondLength.toDouble()).toInt()) {
            val firstChar = first[i].lowercaseChar()
            val secondChar = second[i].lowercaseChar()

            val compare = firstChar.compareTo(secondChar)
            if (compare != 0) {
                return compare
            }
        }

        return firstLength.compareTo(secondLength)
    }
}
